### PR TITLE
Fix load error with generic annotations (#757)

### LINKED ISSUE #757: Load error when declaring a generic mixin on a single class using `@With` annotation
**Describe the bug**

I want to use a generic mixin on a single class within a freezed model. According to [the Readme](https://github.com/rrousselGit/freezed/blob/29aafcc3db2d4ba2f528543fd683beb56f6f7515/packages/freezed/README.md#mixins-and-interfaces-for-individual-classes-for-union-types) we should use `With.fromString` although that doesn't seem to exist when I try to use it and the [Changelog indicates that it was removed in 0.15.0](https://github.com/rrousselGit/freezed/blob/29aafcc3db2d4ba2f528543fd683beb56f6f7515/packages/freezed/CHANGELOG.md#0150). The new syntax generates valid code when build_runner is run but when executing the code the `@With` declaration causes a load error:

```
Error: Type variables can't be used in static members.
```

**To Reproduce**

I've created a [small repo](https://github.com/rorystephenson/freezed-bug-example) with a test case that reproduces the error, the relevant code is also included below:

The mixin:
```dart
mixin ExampleMixin<T> {}
```

The freezed model:
```dart
import 'package:freezed_annotation/freezed_annotation.dart';

import 'example_mixin.dart';

part 'freezed_bug_example.freezed.dart';

@freezed
class FreezedBugExample<T> with _$FreezedBugExample<T> {
  @With<ExampleMixin<T>>()
  factory FreezedBugExample.mixinExample() = MixinExample<T>;

  factory FreezedBugExample.noMixinExample() = NoMixinExample<T>;
}
```

<details>
  <summary>The generated freezed code:</summary>
  
```dart
// coverage:ignore-file
// GENERATED CODE - DO NOT MODIFY BY HAND
// ignore_for_file: type=lint
// ignore_for_file: unused_element, deprecated_member_use, deprecated_member_use_from_same_package, use_function_type_syntax_for_parameters, unnecessary_const, avoid_init_to_null, invalid_override_different_default_values_named, prefer_expression_function_bodies, annotate_overrides, invalid_annotation_target

part of 'freezed_bug_example.dart';

// **************************************************************************
// FreezedGenerator
// **************************************************************************

T _$identity<T>(T value) => value;

final _privateConstructorUsedError = UnsupportedError(
    'It seems like you constructed your class using `MyClass._()`. This constructor is only meant to be used by freezed and you are not supposed to need it nor use it.\nPlease check the documentation here for more information: https://github.com/rrousselGit/freezed#custom-getters-and-methods');

/// @nodoc
mixin _$FreezedBugExample<T> {
  @optionalTypeArgs
  TResult when<TResult extends Object?>({
    required TResult Function() mixinExample,
    required TResult Function() noMixinExample,
  }) =>
      throw _privateConstructorUsedError;
  @optionalTypeArgs
  TResult? whenOrNull<TResult extends Object?>({
    TResult Function()? mixinExample,
    TResult Function()? noMixinExample,
  }) =>
      throw _privateConstructorUsedError;
  @optionalTypeArgs
  TResult maybeWhen<TResult extends Object?>({
    TResult Function()? mixinExample,
    TResult Function()? noMixinExample,
    required TResult orElse(),
  }) =>
      throw _privateConstructorUsedError;
  @optionalTypeArgs
  TResult map<TResult extends Object?>({
    required TResult Function(MixinExample<T> value) mixinExample,
    required TResult Function(NoMixinExample<T> value) noMixinExample,
  }) =>
      throw _privateConstructorUsedError;
  @optionalTypeArgs
  TResult? mapOrNull<TResult extends Object?>({
    TResult Function(MixinExample<T> value)? mixinExample,
    TResult Function(NoMixinExample<T> value)? noMixinExample,
  }) =>
      throw _privateConstructorUsedError;
  @optionalTypeArgs
  TResult maybeMap<TResult extends Object?>({
    TResult Function(MixinExample<T> value)? mixinExample,
    TResult Function(NoMixinExample<T> value)? noMixinExample,
    required TResult orElse(),
  }) =>
      throw _privateConstructorUsedError;
}

/// @nodoc
abstract class $FreezedBugExampleCopyWith<T, $Res> {
  factory $FreezedBugExampleCopyWith(FreezedBugExample<T> value,
          $Res Function(FreezedBugExample<T>) then) =
      _$FreezedBugExampleCopyWithImpl<T, $Res>;
}

/// @nodoc
class _$FreezedBugExampleCopyWithImpl<T, $Res>
    implements $FreezedBugExampleCopyWith<T, $Res> {
  _$FreezedBugExampleCopyWithImpl(this._value, this._then);

  final FreezedBugExample<T> _value;
  // ignore: unused_field
  final $Res Function(FreezedBugExample<T>) _then;
}

/// @nodoc
abstract class _$$MixinExampleCopyWith<T, $Res> {
  factory _$$MixinExampleCopyWith(
          _$MixinExample<T> value, $Res Function(_$MixinExample<T>) then) =
      __$$MixinExampleCopyWithImpl<T, $Res>;
}

/// @nodoc
class __$$MixinExampleCopyWithImpl<T, $Res>
    extends _$FreezedBugExampleCopyWithImpl<T, $Res>
    implements _$$MixinExampleCopyWith<T, $Res> {
  __$$MixinExampleCopyWithImpl(
      _$MixinExample<T> _value, $Res Function(_$MixinExample<T>) _then)
      : super(_value, (v) => _then(v as _$MixinExample<T>));

  @override
  _$MixinExample<T> get _value => super._value as _$MixinExample<T>;
}

/// @nodoc

class _$MixinExample<T> with ExampleMixin<T> implements MixinExample<T> {
  _$MixinExample();

  @override
  String toString() {
    return 'FreezedBugExample<$T>.mixinExample()';
  }

  @override
  bool operator ==(dynamic other) {
    return identical(this, other) ||
        (other.runtimeType == runtimeType && other is _$MixinExample<T>);
  }

  @override
  int get hashCode => runtimeType.hashCode;

  @override
  @optionalTypeArgs
  TResult when<TResult extends Object?>({
    required TResult Function() mixinExample,
    required TResult Function() noMixinExample,
  }) {
    return mixinExample();
  }

  @override
  @optionalTypeArgs
  TResult? whenOrNull<TResult extends Object?>({
    TResult Function()? mixinExample,
    TResult Function()? noMixinExample,
  }) {
    return mixinExample?.call();
  }

  @override
  @optionalTypeArgs
  TResult maybeWhen<TResult extends Object?>({
    TResult Function()? mixinExample,
    TResult Function()? noMixinExample,
    required TResult orElse(),
  }) {
    if (mixinExample != null) {
      return mixinExample();
    }
    return orElse();
  }

  @override
  @optionalTypeArgs
  TResult map<TResult extends Object?>({
    required TResult Function(MixinExample<T> value) mixinExample,
    required TResult Function(NoMixinExample<T> value) noMixinExample,
  }) {
    return mixinExample(this);
  }

  @override
  @optionalTypeArgs
  TResult? mapOrNull<TResult extends Object?>({
    TResult Function(MixinExample<T> value)? mixinExample,
    TResult Function(NoMixinExample<T> value)? noMixinExample,
  }) {
    return mixinExample?.call(this);
  }

  @override
  @optionalTypeArgs
  TResult maybeMap<TResult extends Object?>({
    TResult Function(MixinExample<T> value)? mixinExample,
    TResult Function(NoMixinExample<T> value)? noMixinExample,
    required TResult orElse(),
  }) {
    if (mixinExample != null) {
      return mixinExample(this);
    }
    return orElse();
  }
}

abstract class MixinExample<T>
    implements FreezedBugExample<T>, ExampleMixin<T> {
  factory MixinExample() = _$MixinExample<T>;
}

/// @nodoc
abstract class _$$NoMixinExampleCopyWith<T, $Res> {
  factory _$$NoMixinExampleCopyWith(
          _$NoMixinExample<T> value, $Res Function(_$NoMixinExample<T>) then) =
      __$$NoMixinExampleCopyWithImpl<T, $Res>;
}

/// @nodoc
class __$$NoMixinExampleCopyWithImpl<T, $Res>
    extends _$FreezedBugExampleCopyWithImpl<T, $Res>
    implements _$$NoMixinExampleCopyWith<T, $Res> {
  __$$NoMixinExampleCopyWithImpl(
      _$NoMixinExample<T> _value, $Res Function(_$NoMixinExample<T>) _then)
      : super(_value, (v) => _then(v as _$NoMixinExample<T>));

  @override
  _$NoMixinExample<T> get _value => super._value as _$NoMixinExample<T>;
}

/// @nodoc

class _$NoMixinExample<T> implements NoMixinExample<T> {
  _$NoMixinExample();

  @override
  String toString() {
    return 'FreezedBugExample<$T>.noMixinExample()';
  }

  @override
  bool operator ==(dynamic other) {
    return identical(this, other) ||
        (other.runtimeType == runtimeType && other is _$NoMixinExample<T>);
  }

  @override
  int get hashCode => runtimeType.hashCode;

  @override
  @optionalTypeArgs
  TResult when<TResult extends Object?>({
    required TResult Function() mixinExample,
    required TResult Function() noMixinExample,
  }) {
    return noMixinExample();
  }

  @override
  @optionalTypeArgs
  TResult? whenOrNull<TResult extends Object?>({
    TResult Function()? mixinExample,
    TResult Function()? noMixinExample,
  }) {
    return noMixinExample?.call();
  }

  @override
  @optionalTypeArgs
  TResult maybeWhen<TResult extends Object?>({
    TResult Function()? mixinExample,
    TResult Function()? noMixinExample,
    required TResult orElse(),
  }) {
    if (noMixinExample != null) {
      return noMixinExample();
    }
    return orElse();
  }

  @override
  @optionalTypeArgs
  TResult map<TResult extends Object?>({
    required TResult Function(MixinExample<T> value) mixinExample,
    required TResult Function(NoMixinExample<T> value) noMixinExample,
  }) {
    return noMixinExample(this);
  }

  @override
  @optionalTypeArgs
  TResult? mapOrNull<TResult extends Object?>({
    TResult Function(MixinExample<T> value)? mixinExample,
    TResult Function(NoMixinExample<T> value)? noMixinExample,
  }) {
    return noMixinExample?.call(this);
  }

  @override
  @optionalTypeArgs
  TResult maybeMap<TResult extends Object?>({
    TResult Function(MixinExample<T> value)? mixinExample,
    TResult Function(NoMixinExample<T> value)? noMixinExample,
    required TResult orElse(),
  }) {
    if (noMixinExample != null) {
      return noMixinExample(this);
    }
    return orElse();
  }
}

abstract class NoMixinExample<T> implements FreezedBugExample<T> {
  factory NoMixinExample() = _$NoMixinExample<T>;
}

```

</details>

Test code:

```dart
import 'package:freezed_bug_report/freezed_bug_example.dart';
import 'package:test/test.dart';

void main() {
  test('bug example', () {
    FreezedBugExample.mixinExample();
  });
}
```

Running this results in:

```
Failed to load ".../test/freezed_bug_report_test.dart":
lib/freezed_bug_example.dart:9:22: Error: Type variables can't be used in static members.
  @With<ExampleMixin<T>>()
                     ^
lib/freezed_bug_example.dart:12:22: Error: Type variables can't be used in static members.
  @With<ExampleMixin<T>>()
```

**Expected behavior**

No load error should be thrown. I think this is a result of dropping the `@With.fromString` syntax. If I remove `@With<ExampleMixin<T>>()` and run the code without re-running `build_runner` it loads and executes fine so the generated code seems OK, dart just doesn't like the new way of declaring the mixin with a generic type.

### PR DESCRIPTION
**Pull request by rorystephenson**
Fixes #757 

The generated code seems to be valid and the tests run locally. Let me know what you thing @rrousselGit .

### CODE REVIEW

**Comment by rorystephenson** [comment]
@rrousselGit when I run `flutter format --set-exit-if-changed .` in freezed it makes no changes locally so I'm not quite sure why the check is failing?

**Comment by rrousselGit** [comment]
Hello! Could you write tests for these?

**Comment by rorystephenson** [comment]
> Hello! Could you write tests for these?

I've written one, let me know if it's enough :).

File: packages/freezed/test/implements_decorator_test.dart [hunk 1]
```diff
@@ -62,12 +62,23 @@ void main() {
...
+      const object = GenericImplements<int>.city('Morning', 140000);
+      expect(object, isA<House>());
+      expect(
+        object,
+        isA<GeographicArea>().having((s) => s.name, 'name', 'Morning'),
+      );
+    });
+
+    test('generic @With and @Implements using fromString', () {
+      const object = GenericImplements<int>.street('Oak', 42);
```
Comment (rrousselGit) [inline packages/freezed/test/implements_decorator_test.dart:76]:
```suggestion
      const object = GenericStreet<int>('Oak', 42);
      // testing that GenericStreet's abstract class implements the types too
      AdministrativeArea<int> area = object;
      Generic<int> generic = object;
```

File: packages/freezed/test/implements_decorator_test.dart [hunk 2]
```diff
@@ -62,12 +62,23 @@ void main() {
...
-  test('Generic type', () {
-    const object = GenericImplements<int>.city('Morning', 140000);
-    expect(object, isA<House>());
-    expect(
-      object,
-      isA<GeographicArea>().having((s) => s.name, 'name', 'Morning'),
-    );
+  group('Generic type', () {
+    test('static @With and @Implements', () {
+      const object = GenericImplements<int>.city('Morning', 140000);
```
Comment (rrousselGit) [inline packages/freezed/test/implements_decorator_test.dart:67]:
```suggestion
      const object = GenericCity<int>('Morning', 140000);
      // testing that GenericStreet's abstract class implements the types too
      House house = object;
      GeographicArea area = object;
```

**Comment by rrousselGit** [comment]
Great!

One last thing, could you add docs for this? Both in the readme and in the Dart code

File: packages/freezed/test/implements_decorator_test.dart [hunk 3]
```diff
@@ -62,12 +62,29 @@ void main() {
...
-      object,
-      isA<GeographicArea>().having((s) => s.name, 'name', 'Morning'),
-    );
+  group('Generic type', () {
+    test('static @With and @Implements', () {
+      const object = GenericCity<int>('Morning', 140000);
+      // testing that GenericStreet's abstract class implements the types too
+      House house = object;
+      GeographicArea area = object;
+      expect(object, isA<House>());
```
Comment (rrousselGit) [inline packages/freezed/test/implements_decorator_test.dart:71]:
```suggestion
      // testing that GenericStreet's abstract class implements the types too
      // ignore: unused_local_variable
      House house = object;
      // ignore: unused_local_variable
      GeographicArea area = object;

      expect(object, isA<House>());
```

File: packages/freezed/test/implements_decorator_test.dart [hunk 4]
```diff
@@ -62,12 +62,29 @@ void main() {
...
+        isA<GeographicArea>().having((s) => s.name, 'name', 'Morning'),
+      );
+    });
+
+    test('generic @With and @Implements using fromString', () {
+      const object = GenericStreet<int>('Oak', 42);
+      // testing that GenericStreet's abstract class implements the types too
+      AdministrativeArea<int> area = object;
+      Generic<int> generic = object;
+      expect(object, isA<AdministrativeArea<int>>());
```
Comment (rrousselGit) [inline packages/freezed/test/implements_decorator_test.dart:83]:
```suggestion
      // testing that GenericStreet's abstract class implements the types too
      // ignore: unused_local_variable
      AdministrativeArea<int> area = object;
      // ignore: unused_local_variable
      Generic<int> generic = object;

      expect(object, isA<AdministrativeArea<int>>());
```

**Comment by rorystephenson** [comment]
The README was not updated when fromString was removed so I had left it as is but on reviewing it I decided to clarify when it's necessary and fix the example.

Similarly the annotations already had updated documentation on the classes but not on the constructors so I've added some documentation there too.

**Comment by rrousselGit** [comment]
This is great, thanks!

**Pull request merged**